### PR TITLE
Fix extracting logic in dynamic filtering when integrated with filter pushdown

### PR DESCRIPTION
**Pull request by kewang1024**
Test plan - 
1. Integration Test
2.Test in internal verifier runs

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Review by highker — COMMENTED** [review]
- Could you write a test for this?
- Avoid mentioning word "aria" in the commit title/message

**Review by highker — COMMENTED** [review]
- The test added can pass with or without the fix. Can we have a better test?
- The fix doesn't seem to be at the right place. Let's figure out a proper test first and later come back to the fix.

**Comment by kewang1024** [comment]
Nice catch, I pasted the wrong query.
I produced in another branch and forgot to paste that query here

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveDistributedJoinQueriesWithDynamicFilteringAndFilterPushdown.java [hunk 1]
```diff
@@ -43,4 +44,35 @@ protected Session getSession()
...
+                "l1.orderkey = o1.orderkey AND o1.custkey = 397\n" +
+                ")\n" +
+                ")\n" +
+                "LEFT OUTER JOIN part p2\n" +
+                "ON (\n" +
+                "p1.name = p2.name AND p1.partkey = p2.partkey\n" +
+                ")\n" +
+                ")\n" +
+                "WHERE\n" +
+                "o1.shippriority = 0");
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveDistributedJoinQueriesWithDynamicFilteringAndFilterPushdown.java:76]: Could you make the body of the SQL more readable?
```
SELECT * FROM
lineitem l1 LEFT OUTER JOIN part p1
ON l1.orderkey = p1.partkey AND p1.size = 47
INNER JOIN orders o1 ON l1.orderkey = o1.orderkey
AND o1.custkey = 397
LEFT OUTER JOIN part p2
ON p1.name = p2.name AND p1.partkey = p2.partkey
WHERE o1.shippriority = 0

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveDistributedJoinQueriesWithDynamicFilteringAndFilterPushdown.java [hunk 2]
```diff
@@ -43,4 +44,35 @@ protected Session getSession()
                 .setCatalogSessionProperty(HIVE_CATALOG, PUSHDOWN_FILTER_ENABLED, "true")
                 .build();
     }
+
+    @Test
+    // Mixed join could produce conjunction dynamic filters, we should be able to extract them out when integrating with filter pushdown
+    // In this query, it will produce:
+    // AND (($internal$dynamic_filter_function(VARCHAR 263, partkey)) OR (IS_NULL(partkey))))
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveDistributedJoinQueriesWithDynamicFilteringAndFilterPushdown.java:51]: remove these two lines

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveDistributedJoinQueriesWithDynamicFilteringAndFilterPushdown.java [hunk 3]
```diff
@@ -43,4 +44,35 @@ protected Session getSession()
                 .setCatalogSessionProperty(HIVE_CATALOG, PUSHDOWN_FILTER_ENABLED, "true")
                 .build();
     }
+
+    @Test
+    // Mixed join could produce conjunction dynamic filters, we should be able to extract them out when integrating with filter pushdown
```
Comment (highker) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveDistributedJoinQueriesWithDynamicFilteringAndFilterPushdown.java:49]: move this comment to the body of the test

**Review by highker — COMMENTED** [review]
this patch will take me longer to review. I need to think about how to fix it properly

**Review by highker — COMMENTED** [review]
The root cause of this issue is actual because of we pushing down filters before calling `RemoveUnsupportedDynamicFilters`. Note that your current fix is essentially very similar to `RemoveUnsupportedDynamicFilters.removeNestedDynamicFilters`. So here is what I'm thinking:
- make `removeNestedDynamicFilters` a common utils in `DynamicFilters`
- extract the dynamic filters from the filter to be pushed down as is but apply one more step in `remainingExpression`: apply `removeNestedDynamicFilters` to `remainingExpression`, so that we remove nested dynamic filters to be pushed down.

**Review by highker — APPROVED** [review]

**Pull request merged**